### ISSUE TITLE
I have some questions about the data parse section

### ISSUE THREAD

**Issue by wangwenying1** [alexjamesmx/Deeplog-and-loganomaly#1]
hello!teacher.About the data parsing part of these two papers, I have some questions.
the first one is Deeplog, where the data parsing gets the log key and log template, and then how to get the log key sequence? The second paper Loganomaly, about the data processing part of this paper, I don't know how to implement it. I see that you have preprocessing code here, is it the template2Vec part? (See the relevant code section to experiment directly with the same input as deeplog(HDFS_train), link here https://github.com/d0ng1ee/logdeep?tab=readme-ov-file)

**Comment by alexjamesmx** [comment]
Hi, I'll update the code with HDFS and BGL datasets soon so you can test it. In deeplog, each event template is mapped to a index, so in a window of w logs, instead of showing the eventTemplates it'll show the indices to their respective mappings. Then from all this sequences, the labels are the next index to occur. This sequences and labels are then fed into the model so given a sequence it can predict the next label.

LogAnomaly uses semantics with synonyms and antonyms. However, here the embeddings are just vector representations of the templates. This is cause original paper's model is not public. So when an unseen event appears. it is compared to the already known (trained) templates and if it is very similar then it maps it to the known template so less false positives occur. 

The implementation here is deeplog's log key model, and as add-on, mapping unseen events from logAnomaly to reduce false positives (not neccesary better, cause it might misclassify a real anomaly).

The log key sequence size is up to you, you can use fixed windows, sliding windows or session windows. The time frame that you think your log data can be adapted to. You might also use timestamp and use time.

If you have any further questions, I will be happy to help, although I am not the author of the articles and I think it is best to send any complex clarifications to these people.

**Comment by wangwenying1** [comment]
Thank you very much for your reply, I will take some time to digest it first. 

Then it occurred to me that deeplog's paper has been puzzling me. Regarding the log key anomaly detection, I obtained structured logs and template logs using drain parse. Then how should I get the log sequence? Is it according to the same BLK_5556055514732131855 classification, each blk is sorted according to the timestamp of the log key to get the log sequence, my understanding is correct?

The second is your code, what is the execution order? The data set used here was made using log data that you performed yourself, right?

**Comment by alexjamesmx** [comment]
Yeah Im using a private dataset but this code was implemented with hdfs, bgl and spirit dataset before and gave similar results to the ones claimed. Part of this code is from LogADempirical and semPCA which are 2 interesting repos you could see.

So look, HDFS is a dataset that each Block id represents sort of allocation, deletion, and other processes. In HDFS especifically, in most papers and codes, using session_window gives better results than slidings or fixed. Due to dataset nature. So the HDFS sequences are created by gathering all logs per block_id, then this groups are split into w windows. 

So you'll have something like this for example
this session containing all these events which also represent eventTemplates. 
 
sequence/index 1: {'SessionId': 'blk_-6531739066606093691', 'EventId': ['bbb51b95', 'bbb51b95', 'bbb51b95', '3d91fa85', 'd38aa58d'], 'EventTemplate': ['Receiving block <*> src: /<*> dest: /<*>', 'Receiving block <*> src: /<*> dest: /<*>', 'Receiving block <*> src: /<*> dest: /<*>', 'BLOCK* NameSystem.allocateBlock: <*> <*>', 'PacketResponder <*> for block <*> <*>'], 'Label': 0} 

This is a sequence where the label is 0.

Now, if drain parser gave you the templates it'll look something like this: 
padding 11
BLOCK* NameSystem.allocateBlock: <*> <*> 1
Received block <*> src: /<*> dest: /<*> of size <*> 2
<*> Starting thread to transfer block <*> to <*> 3
PacketResponder <*> for block <*> <*> 4
Unexpected error trying to delete block <*>. BlockInfo not found in volumeMap. 5
Receiving block <*> src: /<*> dest: /<*> 6
Verification succeeded for <*> 7
BLOCK* ask <*> to replicate <*> to datanode(s) <*> 8
BLOCK* NameSystem.addStoredBlock: blockMap updated: <*> is added to <*> size <*> 9
Received block <*> of size <*> from /<*> 10
<*>:Got exception while serving <*> to /<*>: 12
BLOCK* NameSystem.delete: <*> is added to invalidSet of <*> 13
<*>:Transmitted block <*> to /<*> 14
Deleting block <*> file <*> 15
<*> Served block <*> to /<*> 16


So now each eventTemplate in each sequence is mapped to a number which may be different for you.
so the sequence would finally look like this: 
 {'SessionId': 'blk_-6531739066606093691', 'EventId': ['bbb51b95', 'bbb51b95', 'bbb51b95', '3d91fa85', 'd38aa58d'], 'EventTemplate': [6, 6, 6 ,1 , 4], 'Label': 0} 


that'll be an example of a final sequence to then feed the lstm model
